components/proxy/src/test/kotlin/com/hotels/styx/routing/handlers/LoadBalancingGroupTest.kt

```diff
@@ -0,0 +1,218 @@
+import com.hotels.styx.api.configuration.ObjectStore
+import com.hotels.styx.api.exceptions.NoAvailableHostsException
+import com.hotels.styx.routing.RoutingContext
+import com.hotels.styx.routing.RoutingObjectRecord
+import com.hotels.styx.routing.db.StyxObjectStore
+import com.hotels.styx.routing.handle
+import com.hotels.styx.routing.routingObjectDef
+import com.hotels.styx.server.HttpInterceptorContext
+import io.kotlintest.IsolationMode
+import io.kotlintest.matchers.numerics.shouldBeGreaterThan
+import io.kotlintest.matchers.types.shouldBeNull
+import io.kotlintest.shouldBe
+import io.kotlintest.shouldThrow
+import io.kotlintest.specs.FeatureSpec
+import io.mockk.mockk
+import io.mockk.verify
+import org.reactivestreams.Publisher
+import reactor.core.Disposable
+import reactor.core.publisher.toFlux
+import reactor.core.publisher.toMono
+import java.nio.charset.StandardCharsets.UTF_8
+import java.time.Duration
+
+class LoadBalancingGroupTest : FeatureSpec() {
+
+    // Tests depend on each other. So run the tests sequentially:
+    override fun isolationMode(): IsolationMode = IsolationMode.SingleInstance
+
+    init {
+        feature("Load Balancing") {
+            val factory = LoadBalancingGroup.Factory()
+            val routeDb = StyxObjectStore<RoutingObjectRecord>()
+
+            routeDb.insert("appx-01", RoutingObjectRecord.create("HostProxy", setOf("appX"), mockk(), StaticResponseHandler(200, "appx-01")))
+            routeDb.insert("appx-02", RoutingObjectRecord.create("HostProxy", setOf("appX"), mockk(), StaticResponseHandler(200, "appx-02")))
+            routeDb.insert("appx-03", RoutingObjectRecord.create("HostProxy", setOf("appX"), mockk(), StaticResponseHandler(200, "appx-03")))
+
+            routeDb.insert("appy-01", RoutingObjectRecord.create("HostProxy", setOf("appY"), mockk(), StaticResponseHandler(200, "appy-01")))
+            routeDb.insert("appy-02", RoutingObjectRecord.create("HostProxy", setOf("appY"), mockk(), StaticResponseHandler(200, "appy-02")))
+
+            routeDb.watch().waitUntil { it.entrySet().size == 5 }
+
+            val lbGroup = factory.build(listOf("appX"), RoutingContext(routeDb = routeDb).get(), routingObjectDef("""
+                type: LoadBalancingGroup
+                config:
+                  origins: appX
+                """.trimIndent())) as LoadBalancingGroup
+
+            scenario("Discovers origins with appropriate tag") {
+                val frequencies = mutableMapOf<String, Int>()
+
+                for (i in 1..100) {
+                    lbGroup.call(get("/").build())
+                            .bodyAs(UTF_8)
+                            .let {
+                                val current = frequencies.getOrDefault(it, 0)
+                                frequencies[it] = current + 1
+                            }
+                }
+
+                println("Frequencies: " + frequencies)
+
+                frequencies["appx-01"]?.shouldBeGreaterThan(20)
```

Comment:
Is this a special case where a method can be called on `null`?

Reply:
Please clarify. What method are you referring to? The get operator, or `shouldBeGreagerThan` or something else?

Reply:
I'm confused about the use of `?.` - normally this would only call `shouldBeGreaterThan` if `frequencies["appx-01"]` were non-null, meaning it would pass the test by being null (which is not the expected behaviour), so I would expect `!!` to be used here instead.

However, further down I see `frequencies["appy-01"]?.shouldBeNull()` - and the method `shouldBeNull` could only have a reason to exist if can somehow be called on null values (though in that case wouldn't the `?.` be better replaced with just `.`?)

So it seems like a confusing feature of the test framework, as it leaves me uncertain about what I should expect.

Reply:
> I'm confused about the use of ?. - normally this would only call shouldBeGreaterThan if frequencies["appx-01"] were non-null, meaning it would pass the test by being null (which is not the expected behaviour), so I would expect !! to be used here instead.

Good point. This is clearly a Kotlin newbie error! This actually skips the validation altogether for null values! I'll review all other usages of this.

> However, further down I see frequencies["appy-01"]?.shouldBeNull() - and the method shouldBeNull could only have a reason to exist if can somehow be called on null values (though in that case wouldn't the ?. be better replaced with just .?)

Here you need to realise that `shouldBeNull` is not a method. It is an extension function on `Any?`. See: https://kotlinlang.org/docs/reference/extensions.html. That is, the value `frequencies["apply-01"]` is passed to `shouldBeNull` extension function that just asserts it is null. Therefore the safe call operator is totally unnecessary.